### PR TITLE
added RemoteWorkspace

### PR DESCRIPTION
**Pull request by heejaechang**
RemoteWorkspace has host agnostic implementation of roslyn features/services/workspace that will run in remote host

commit with actual changes
https://github.com/dotnet/roslyn/commit/7e05de874a524ecbb6553ad7e7136b1b9fa9d0aa

work left for remote workspaces
https://github.com/dotnet/roslyn/issues/12776

### CODE REVIEW

**Comment by heejaechang** [comment]
@mavasani @mattwar @KevinH-MS @dotnet/roslyn-ide @CyrusNajmabadi @panopticoncentral can you take a look?

**Comment by panopticoncentral** [comment]
:+1:

File: src/Workspaces/Remote/Core/Services/AssetService.cs [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+
+        private readonly ConcurrentDictionary<int, AssetSource> _assetSources =
+            new ConcurrentDictionary<int, AssetSource>(concurrencyLevel: 4, capacity: 10);
+
+        private readonly ConcurrentDictionary<Checksum, Entry> _assets =
+            new ConcurrentDictionary<Checksum, Entry>(concurrencyLevel: 4, capacity: 10);
+
+        public AssetService()
+        {
+            Task.Factory.StartNew(CleanAssets, CancellationToken.None, TaskCreationOptions.LongRunning, TaskScheduler.Default).Unwrap();
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/Core/Services/AssetService.cs]: what is the unwrap for?

Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/Core/Services/AssetService.cs]: Why is this not just Task.Run?

File: src/Workspaces/Remote/Core/Services/AssetService.cs [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+namespace Microsoft.CodeAnalysis.Remote
+{
+    /// <summary>
+    /// This service provide a way to get roslyn objects from checksum
+    /// 
+    /// TODO: change this service to workspace service
+    /// </summary>
+    internal class AssetService
+    {
+        private const int CleanupInterval = 3; // 3 minutes
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/Core/Services/AssetService.cs]: 3 minutes per asset seems like a long time.  If, for example, the user is heavily editing a single large file, won't this just cause a lot to be kept around?

File: src/Workspaces/Remote/Core/Services/AssetService.cs [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+
+        public async Task<T> GetAssetAsync<T>(Checksum checksum, CancellationToken cancellationToken)
+        {
+            Entry entry;
+            if (!_assets.TryGetValue(checksum, out entry))
+            {
+                // TODO: what happen if service doesn't come back. timeout?
+                await RequestAssetAsync(checksum, cancellationToken).ConfigureAwait(false);
+
+                if (!_assets.TryGetValue(checksum, out entry))
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/Core/Services/AssetService.cs]: It feels like there's a race here.  Someone requests the asset, aand you call into RequestAssetAsync.  Then between that line and ``` if (!_assets.TryGetValue(checksum, out entry))```, you have the code that purges assets.

File: src/Workspaces/Remote/Core/Services/AssetService.cs [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+                try
+                {
+                    // ask one of asset source for data
+                    await source.RequestAssetAsync(serviceId, checksum, cancellationToken).ConfigureAwait(false);
+                }
+                catch (Exception ex)
+                {
+                    // connection to the asset source has closed.
+                    // move to next asset source
+                    Contract.ThrowIfFalse(ex is OperationCanceledException || ex is IOException || ex is ObjectDisposedException);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/Core/Services/AssetService.cs]: just catch these specific exceptions.  or use a catch filter.

File: src/Workspaces/Remote/Core/Services/AssetService.cs [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+                    _assets.TryRemove(kvp.Key, out entry);
+                }
+
+                await Task.Delay(cleanupIntervalTimeSpan).ConfigureAwait(false);
+            }
+        }
+
+        public void Set(Checksum checksum, object value)
+        {
+            _assets.TryAdd(checksum, new Entry(value));
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/Core/Services/AssetService.cs]: Having external people able to set assets here seems wonky.  Why doesn't RequestAssetAsync return the asset and this type can then put it in its cache itself.

File: src/Workspaces/Remote/Core/Services/AssetService.cs (hunk 2 above)
Comment (heejaechang) [inline src/Workspaces/Remote/Core/Services/AssetService.cs]: yep. need better caching mechanism. already have tracking bug.

File: src/Workspaces/Remote/Core/Services/AssetService.cs (hunk 1 above)
Comment (heejaechang) [inline src/Workspaces/Remote/Core/Services/AssetService.cs]: since we don't hold onto returned task, unwrap doesn't needed.

Task.Run doesn't support creation option such as LongRunning.

File: src/Workspaces/Remote/Core/Services/AssetService.cs (hunk 5 above)
Comment (heejaechang) [inline src/Workspaces/Remote/Core/Services/AssetService.cs]: sure, we can do that way.

File: src/Workspaces/Remote/Core/Services/AssetService.cs (hunk 3 above)
Comment (heejaechang) [inline src/Workspaces/Remote/Core/Services/AssetService.cs]: you mean request asset can take 30 min?

File: src/Workspaces/Remote/Core/Services/AssetService.cs (hunk 4 above)
Comment (heejaechang) [inline src/Workspaces/Remote/Core/Services/AssetService.cs]: sure.

File: src/Workspaces/Remote/Core/Services/AssetService.cs (hunk 1 above)
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/Core/Services/AssetService.cs]: Task.LongRunning shouldn't be necessary here.  When your task does run it will be for very short periods, and will relinquish control again almost immediately when it does the 'await delay'.  LongRunning is when you expect your task to need to occupy the use of the thread for significant periods of time.

File: src/Workspaces/Remote/Core/Services/AssetService.cs (hunk 3 above)
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/Core/Services/AssetService.cs]: I simply mean that there's a race in this code.  You have two overlapping operations.  One checks a cache, requests an item (which will put the item into the cache) and checks the cache again.  The other is going through the cache and deleting things.

I really don't like that.  I'd far prefer that hte request of the asset returns the data.  You can then know that you have it without having to have scary Contract.Fail calls.

In other words: design so that it can't fail, instead of having to hope that it won't fail.

File: src/Workspaces/Remote/Core/Services/AssetService.cs (hunk 1 above)
Comment (heejaechang) [inline src/Workspaces/Remote/Core/Services/AssetService.cs]: ya, good point

File: src/Workspaces/Remote/Core/Services/AssetService.cs (hunk 3 above)
Comment (heejaechang) [inline src/Workspaces/Remote/Core/Services/AssetService.cs]: yep. already changed that way. the code has that way since at the beginning, I didn't have my custom pipe stream so, only way to get asset was client calling another service that is not related to this. when I added pipe stream I didn't change the structure. that is why it became this shape.

anyway!. sure.

**Comment by heejaechang** [comment]
this PR is merged to this
https://github.com/dotnet/roslyn/pull/12774
